samcli/local/docker/container.py

```diff
@@ -272,6 +277,35 @@ def start(self, input_data=None):
         # Start the container
         real_container.start()
 
+        # Wait for port to be open
+        self.wait_for_port()
+
+    def wait_for_port(self):
+        """
+        Waits until the host machine port that Docker binds to is open.
+        """
+        start_time = time.time()
+        sleep = 0.1
+        while True:
+            a_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            location = (self._container_host_interface, self.rapid_port_host)
+            # connect_ex returns 0 if connection succeeded
+            is_port_open = not a_socket.connect_ex(location)
+            a_socket.close()
+
+            if is_port_open:
+                break
+
+            current_time = time.time()
+            if current_time - start_time > START_CONTAINER_TIMEOUT:
+                raise ContainerNotStartableException(
+                    f"Timed out while starting container. You can increase this timeout by "
+                    f"setting the SAM_CLI_START_CONTAINER_TIMEOUT environment variable. "
+                    f"The current value is {START_CONTAINER_TIMEOUT} (seconds)."
```

Comment:
Discussed this with @moelasmar and even though this message is not 100% accurate (we don't time the whole container start process, but rather just waiting for the port), we thought maybe it's more valuable to have a clear error message.